quickwit/quickwit-ingest/src/ingest_v2/fetch.rs

```diff
@@ -423,79 +359,215 @@ async fn fault_tolerant_fetch_task(
     index_uid: IndexUid,
     source_id: SourceId,
     shard_id: ShardId,
-    mut from_position_exclusive: Option<u64>,
-    to_position_inclusive: Option<u64>,
-
-    preferred_ingester_id: NodeId,
-    mut failover_ingester_id: Option<NodeId>,
-
+    mut from_position_exclusive: Position,
+    to_position_inclusive: Position,
+    ingester_ids: Vec<NodeId>,
     ingester_pool: IngesterPool,
-
-    mut fetch_stream: ServiceStream<IngestV2Result<FetchResponseV2>>,
     fetch_response_tx: mpsc::Sender<IngestV2Result<FetchResponseV2>>,
 ) {
-    while let Some(fetch_response_result) = fetch_stream.next().await {
-        match fetch_response_result {
-            Ok(fetch_response) => {
-                from_position_exclusive = fetch_response.to_position_inclusive();
-                if fetch_response_tx.send(Ok(fetch_response)).await.is_err() {
-                    // The stream was dropped.
-                    break;
+    'outer: for (ingester_idx, ingester_id) in ingester_ids.iter().enumerate() {
+        let failover_ingester_id_opt = ingester_ids.get(ingester_idx + 1);
+
+        let mut ingester = match ingester_pool.get(ingester_id) {
+            Some(ingester) => ingester,
+            _ => {
+                if let Some(failover_ingester_id) = failover_ingester_id_opt {
+                    warn!(
+                        client_id=%client_id,
+                        index_uid=%index_uid,
+                        source_id=%source_id,
+                        shard_id=%shard_id,
+                        "ingester `{ingester_id}` is not available: failing over to ingester `{failover_ingester_id}`"
+                    );
+                } else {
+                    error!(
+                        client_id=%client_id,
+                        index_uid=%index_uid,
+                        source_id=%source_id,
+                        shard_id=%shard_id,
+                        "ingester `{ingester_id}` is not available: closing fetch stream"
+                    );
+                    let ingest_error = IngestV2Error::IngesterUnavailable {
+                        ingester_id: ingester_id.clone(),
+                    };
+                    let _ = fetch_response_tx.send(Err(ingest_error)).await;
```

Comment:
should we break here?

Reply:
Yes. This is our last attempt before giving up and returning. I added a comment.